src/IrcConnection/IrcConnection.cs

```diff
@@ -1470,50 +1206,81 @@ public void Start()
                         
                         // only send active pings if we are registered
                         if (!_Connection.IsRegistered) {
                             continue;
                         }
-                        
+
                         DateTime now = DateTime.Now;
-                        int last_ping_sent = (int)(now - _Connection._LastPingSent).TotalSeconds;
-                        int last_pong_rcvd = (int)(now - _Connection._LastPongReceived).TotalSeconds;
-                        // determins if the resoponse time is ok
-                        if (last_ping_sent < _Connection._PingTimeout) {
-                            if (_Connection._LastPingSent > _Connection._LastPongReceived) {
-                                // there is a pending ping request, we have to wait
-                                continue;
+
+                        // Don't start counting time or sending ping requests until we are registered
+                        if (!wasRegistered) {
+                            _Connection._LastPingSent = now - TimeSpan.FromSeconds(1);
+                            _Connection._LastPongReceived = now;
+                            wasRegistered = true;
+                        }
+
+                        bool timeout = false;
+
+                        // We haven't received a pong since our last ping
+                        // The "=" is critical here.
+                        if (_Connection._LastPongReceived <= _Connection._LastPingSent) {
+
+                            // Has it been long enough to trigger a ping timeout?
+                            if (now.Subtract(_Connection._LastPingSent).TotalSeconds >= _Connection._PingTimeout) {
+                                timeout = true;
                             }
-                            
-                            // determines if it need to send another ping yet
-                            if (last_pong_rcvd > _Connection._PingInterval) {
-                                _Connection.WriteLine(Rfc2812.Ping(_Connection.Address), Priority.Critical);
-                                _Connection._LastPingSent = now;
-                                //_Connection._LastPongReceived = now;
-                            } // else connection is fine, just continue
+
+                        // We have received a pong since our last ping
                         } else {
+
+                            // Is it time to send a new ping?
+                            if (now.Subtract(_Connection._LastPongReceived).TotalSeconds >= _Connection.PingInterval) {
+                                _Connection.WriteLine(Rfc2812.Ping(pingAddress), Priority.Critical);
+                                _Connection._LastPingSent = now;
+                            }
+                        }
+
+                        // Ping timeout
+                        if (timeout) {
+                            System.Diagnostics.Debug.WriteLine("Ping timeout");
```

Comment:
Fixed in IrcConnection and IrcTcpTransport. Where does the logger output go by default?